goal_src/engine/gfx/water/water.gc

```diff
@@ -5,3 +5,2312 @@
+      (logior! (-> s5-0 flags) 2)
+      (set! (-> s5-0 base-height) (-> obj water-height))
+      (set! (-> s5-0 ocean-offset) 0.0)
+      (logior! (-> s5-0 flags) (-> obj flags))
+      (if (< 0.0 (-> obj wade-height))
+       (set! (-> s5-0 wade-height) (-> obj wade-height))
+       )
+      (if (< 0.0 (-> obj swim-height))
+       (set! (-> s5-0 swim-height) (-> obj swim-height))
+       )
+      (if (< 0.0 (-> obj bottom-height))
+       (set! (-> s5-0 bottom-height) (-> obj bottom-height))
+       )
+      (set-zero! (-> s5-0 bob))
+      )
+     )
+    )
+   0
+   (none)
+   )
+  )
+
+(defstate water-vol-startup (water-vol)
+  :virtual #t
+  :code
+  (behavior ()
+   (go-virtual water-vol-idle)
+   (none)
+   )
+  )
+
+(defstate water-vol-idle (water-vol)
+  :virtual #t
+  :event
+  (behavior ((arg0 process) (arg1 int) (arg2 symbol) (arg3 event-message-block))
+   (let ((v1-0 arg2))
+    (the-as object (if (= v1-0 'update)
+                    (TODO-RENAME-26 self)
+                    )
+     )
+    )
+   )
+  :exit
+  (behavior ()
+   (dummy-27 self)
+   (none)
+   )
+  :trans
+  (behavior ()
+   (TODO-RENAME-26 self)
+   (none)
+   )
+  :code
+  (the-as (function none :behavior water-vol) anim-loop)
+  )
+
+(defmethod set-stack-size! water-vol ((obj water-vol))
+  (stack-size-set! (-> obj main-thread) 128)
+  (none)
+  )
+
+(defmethod reset-root! water-vol ((obj water-vol))
+  (set! (-> obj root) (new 'process 'trsqv))
+  (none)
+  )
+
+(defmethod TODO-RENAME-25 water-vol ((obj water-vol))
+  (the-as object 0)
+  )
+
+;; WARN: disable def twice: 85. This may happen when a cond (no else) is nested inside of another conditional, but it should be rare.
+(defmethod TODO-RENAME-29 water-vol ((obj water-vol))
+  (local-vars (r0-0 uint128) (a0-9 uint128) (a0-15 uint128) (sv-16 int))
+  (set!
+   (-> obj attack-event)
+   (the-as
+    symbol
+    ((method-of-type res-lump get-property-struct)
+     (-> obj entity)
+     'attack-event
+     'interp
+     -1000000000.0
+     'drown
+     (the-as (pointer res-tag) #f)
+     *res-static-buf*
+     )
+    )
+   )
+  (process-drawable-from-entity! obj (the-as entity-actor (-> obj entity)))
+  (logclear! (-> obj mask) (process-mask actor-pause))
+  (set! (-> obj vol) (new 'process 'vol-control obj))
+  (logior! (-> obj vol flags) 3)
+  (set! (-> obj bottom-height) 32768.0)
+  (set! (-> obj target) (the-as handle #f))
+  (set! sv-16 0)
+  (let
+   ((v1-8
+     (the-as
+      (pointer float)
+      ((method-of-type res-lump get-property-data)
+       (-> obj entity)
+       'water-height
+       'exact
+       -1000000000.0
+       (the-as pointer #f)
+       (the-as (pointer res-tag) (& sv-16))
+       *res-static-buf*
+       )
+      )
+     )
+    )
+   (when v1-8
+    (set! (-> obj water-height) (-> v1-8 0))
+    (set! (-> obj wade-height) (-> v1-8 1))
+    (set! (-> obj swim-height) (-> v1-8 2))
+    (let ((a0-8 (the-as uint128 sv-16)))
+     (.pcpyud a0-9 a0-8 r0-0)
```

Comment:
this (and others using `r0`) should hopefully disappear once sv-16 is a `res-tag`